rust/src/sdk/api.rs

```diff
@@ -0,0 +1,448 @@
+                ));
+
+                if let Some(model) = global_model {
+                    // global model available
+                    let len = model.len() as c_ulong;
+                    if client.model.is_none() {
+                        // cache the primitive model if needed
+                        client.model = model
+                            .into_primitives()
+                            .map(|res| res.map_err(|_| ()))
+                            .collect::<Result<Vec<$prim_rust>, ()>>()
+                            .map_or(None, |vec| Some(PrimitiveModel::[<$prim_rust:upper>](vec)));
+                    }
+
+                    if let Some(PrimitiveModel::[<$prim_rust:upper>](ref mut model)) = client.model {
+                        // conversion succeeded
+                        let ptr = model.as_mut_ptr() as *mut $prim_c;
+                        [<PrimitiveModel $prim_rust:upper>] { ptr, len }
+                    } else {
+                        // conversion failed
+                        let ptr = ptr::null_mut() as *mut $prim_c;
+                        [<PrimitiveModel $prim_rust:upper>] { ptr, len }
+                    }
+
+                } else {
+                    // global model unavailable
+                    let ptr = ptr::null_mut() as *mut $prim_c;
+                    let len = 0_u64 as c_ulong;
+                    [<PrimitiveModel $prim_rust:upper>] { ptr, len }
+                }
+            }
+        }
+    };
+}
+
+get_model!(f32, c_float);
+get_model!(f64, c_double);
+get_model!(i32, c_int);
+get_model!(i64, c_long);
+
+/// Generates a method to register the updated local model. The argument `$prim` is the
+/// corresponding Rust primitive data type.
+macro_rules! update_model {
+    ($prim:ty $(,)?) => {
+        paste::item! {
+            #[no_mangle]
+            /// Registers the updated local model.
+            ///
+            /// # Safety
+            /// The method dereferences from the raw pointer arguments. Therefore, the behavior of
+            /// the method is undefined if the arguments don't point to valid objects.
+            ///
+            /// The `model` points to memory which is either allocated by `get_model()` and then
+            /// modified or which isn't allocated by `get_model()`. Therefore, the behavior of the
+            /// method is undefined if any of the [slice safety conditions](https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html#safety)
+            /// are violated.
+            pub unsafe extern "C" fn [<update_model_ $prim>](
+                client: *mut Client,
+                model: [<PrimitiveModel $prim:upper>],
+            ) {
+                if !client.is_null()
+                    && !model.ptr.is_null()
+                    && model.len != 0
+                    && model.len <= usize::MAX as c_ulong
+                    && model.len as usize * mem::size_of::<$prim>() <= usize::MAX
```

Comment:
This can overflow if `len` is set to an absurd value.

Reply:
i'll merge line 306 and 307 to
```
&& model.len <= (usize::MAX / mem::size_of::<$prim>()) as c_ulong
```
line 306 was only meant as a pre-check for 307, but as you noted it is necessary but not sufficient.